Friends app

SamVal007 opened the pull request:
#Friends_APP

[demo](https://samval007.github.io/kottans-frontend/Friends_APP/)
[codebase](https://github.com/SamVal007/kottans-frontend/tree/main/Friends_APP)


Please, review.

SamVal007 committed:
Friends_APP

SamVal007 committed:
Friends_APP

zonzujiro commented on submissions/SamVal007/Friends_APP/script.js:
```diff
@@ -0,0 +1,100 @@
+      }
+    )
+    .then(data => data.json())
+    .then(function (data) {
+      persons = data.results;
+      displayList = persons;
+      displayCards();
```
Single responsibility - your functions should do only one job. As an example function in which you fetch users should only fetch them and not render, transform or process them in other ways. All these things should be done in another place, outside your function. The same applied to the sort function, which usually does all types of sorting 😉

SamVal007 replied:
I would be honest, but i didnt find the way how to delete this displayCards() from this piece of code, so i just renamed function. I was trying to put this  displayCards() in document.addEventListener('DOMContentLoaded', function () , but it didnt work...

zonzujiro replied:
Tip - you should use `then` in special place ;)

SamVal007 replied:
Yep, this I understood, but where:)
Ok, I will try to figure out!

zonzujiro replied:
"in special place" 😄

SamVal007 replied:
Ok, I think I found out how to do it!
First of all - I tried to use one more **promise** inside of  `document.addEventListener('DOMContentLoaded', function (){...}` , but it didnt work, i dont know why. Then I decided to use `async` and seems it worked!

zonzujiro commented on submissions/SamVal007/Friends_APP/script.js:
```diff
@@ -0,0 +1,100 @@
+          displayList = displayList.sort((a, b) => a.dob.age - b.dob.age);
+          break;
+        case 'ageHigh':
+          displayList = displayList.sort((b, a) => a.dob.age - b.dob.age);
```
1. Why you need assigning to `displayList`?
2. Same function as at line 81. DRY

zonzujiro replied:
Also, please check this comment https://github.com/kottans/frontend-2021-homeworks/pull/449#discussion_r578247714

SamVal007 replied:
Thank you!
Yes, really, i dont need this 'assigning '!
Ok, I will read this comments.

SamVal007 replied:
I'm sorry, Did I understand correctly -  I should  remake this part of code (this switch is too long and it's not "Single responsibility"):

  if (radio) {
      switch (radio.id) {
        case 'genderAll':
          displayList = persons;
          break;
        case 'genderMale':
          displayList = persons.filter(el => el.gender === 'male');
          break;
        case 'genderFemale':
          displayList = persons.filter(el => el.gender === 'female');
          break;
        case 'nameAsc':
          displayList.sort((b, a) => a.name.last > b.name.last ? -1 : 1);
          break;
        case 'nameDesc':
          displayList.sort((b, a) => a.name.last < b.name.last ? -1 : 1);
          break;
        case 'ageLow':
          displayList.sort((a, b) => a.dob.age - b.dob.age);
          break;
        case 'ageHigh':
          displayList.sort((b, a) => a.dob.age - b.dob.age);
          break;
      }

With adding functions for sorting.
BTW, Should I delete "long switch" and make few different "addEventListneres"   for each case of switch or it will be very ig load on site? (sorry may be for stupid questions, but i just want to understand)

zonzujiro replied:
1. Yes, you should delete the long switch
2. You can have a listener on the form and listen to `change` event. Or you can separate listeners between fieldsets. Up to you

SamVal007 replied:
Huh! Unbelievable, but seems I did it!:)
I dodnt know why, but this refactoring took me quit long time. 
Basically, I think. It because I start fixing everything at one time... it was big mistake -  nothing work and could understand why. So, then I started do it step-by-step, like You teaching us:) Refactored sort logic, then long switch. 
Sorry, may be It's not necessary to write it here, but whom else can i write it;))

Ok, so now it's seems

SamVal007 committed:
Update script.js

OleksiyRudenko commented:
@SamVal007 if you think your changes address the reviewer's comments please re-request review

SamVal007 commented:
I'm sorry that I'm out of deadline.
If you would be so kind, re-review it again, please .

SamVal007 committed:
Made changes in order with review

- Add functions for sorting (more structural logic for sorting)
- Changed long switch to more compact logic  (with using addEventListener('change', ({target: radioButton}))

zonzujiro commented on submissions/SamVal007/Friends_APP/script.js:
```diff
@@ -0,0 +1,127 @@
+  nameDesc: () => persons.sort((a, b) => compareName(b, a)),
+  nameAsc: () => persons.sort(compareName)
+};
+const filtrByMale = () => {
```
Just filter by gender and pass `gender` as an argument

zonzujiro commented on submissions/SamVal007/Friends_APP/script.js:
```diff
@@ -0,0 +1,127 @@
+const filtrByFemale = () => {
+  displayList = persons.filter(el => el.gender === 'female');
+};
+//////////////////////////////////////////////////////////////
```
Redundant comments - don’t use stuff ‘———————————‘ or “here I will sort my array”. The first one is ugly, second - redundant because the reviewer will understand that you do sort, from your code. The only situation in which you should use comments is when you need to describe why something is going on and this is 1% of all situations :)

zonzujiro commented on submissions/SamVal007/Friends_APP/script.js:
```diff
@@ -0,0 +1,127 @@
+
+document.addEventListener('DOMContentLoaded', function () {
+  makeFriendList();
+  //displayCards(); //I dodnt know why, but if put displayCards() here -> it doesnt work
```
So you need to found out :D Or you need my help?

SamVal007 replied:
It seems, now it works as we want with single responsibility!
But I watched one video about **pure functions** and understood, that this my code is smelling a little)) Right?:) 
I should make it without using GLOBAL arrays ` let persons = []; let displayList = []; `. Right?

zonzujiro commented on submissions/SamVal007/Friends_APP/script.js:
```diff
@@ -0,0 +1,127 @@
+
+  MENU_FORMS.addEventListener('change', ({target: radioButton}) => {
+    putPersonsInDisplayList();
+    console.log(radioButton);
+    console.log(radioButton.name);
+    console.log(radioButton.id);
```
Please remove `console.log` from PR :)

zonzujiro commented on submissions/SamVal007/Friends_APP/script.js:
```diff
@@ -0,0 +1,127 @@
+    console.log(radioButton.name);
+    console.log(radioButton.id);
+    if (radioButton.name === 'genderField') {
+      if (radioButton.id === 'genderAll') putPersonsInDisplayList();
```
If you are using `radio`, just use `value` on it. Which should be equal to just `female`, `male`, etc.

SamVal007 replied:
Yeap! Cool idea!:) Tnx!

zonzujiro commented on submissions/SamVal007/Friends_APP/script.js:
```diff
@@ -0,0 +1,127 @@
+      if (radioButton.id === 'genderMale') filtrByMale();
+      if (radioButton.id === 'genderFemale') filtrByFemale();
+    }
+    if (radioButton.name === "nameField") {
```
No need for this `Field` part :) It is understandable that you getting value from the input.

zonzujiro commented on submissions/SamVal007/Friends_APP/script.js:
```diff
@@ -0,0 +1,127 @@
+    displayCards();
+  })
+
+  btn.addEventListener('click', () =>{
```
Dnt us shrtcts n nms :)

SamVal007 replied:
Didnt get, what?:)
"Dont use short carts for names"? Right?

zonzujiro replied:
Now you understand

SamVal007 replied:
Ok, Thanks!:)

SamVal007 replied:
@zonzujiro 
What do you think about other places? Is there everything good?

zonzujiro commented on submissions/SamVal007/Friends_APP/script.js:
```diff
@@ -0,0 +1,127 @@
+
+let persons = [];
+let displayList = [];
+const putPersonsInDisplayList = () => {displayList = persons;};
```
Function redundant, IMO

zonzujiro commented on submissions/SamVal007/Friends_APP/script.js:
```diff
@@ -0,0 +1,127 @@
+const putPersonsInDisplayList = () => {displayList = persons;};
+
+function makeFriendList() {
+  let response = fetch(URL)
```
Fetching of data and handling of requests should be in a separate function.

SamVal007 committed:
Update script.js

Made fixes by review.

SamVal007 committed:
Update index.html

SamVal007 committed:
Update script.js

zonzujiro commented:
Sorry, man. I'm with covid right now.

I will return to PR in a week or so.

If not - pls ping me in Telegram :)

SamVal007 commented:
OMG, I'm sorry, I didn't know.
Wish You become Healthy as fast as it is possible.

PS. No worries about PR! It doesn't matter right now!

On Sun, Apr 11, 2021 at 3:50 PM Ivan Tytarenko ***@***.***>
wrote:

> Sorry, man. I'm with covid right now.
>
> I will return to PR in a week or so.
>
> If not - pls ping me in Telegram :)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kottans/frontend-2021-homeworks/pull/457#issuecomment-817302865>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ASDPGHGSXMY6P5L6HXZ56TDTIGLJ7ANCNFSM4XY4JNTQ>
> .
>


-- 
С ув. Самойленко В.

zonzujiro commented on submissions/SamVal007/Friends_APP/script.js:
```diff
@@ -0,0 +1,125 @@
+
+
+const filtrByGender = (value) => {
+  if (value === 'genderAll') putPersonsInDisplayList()
+  else{
+  displayList = persons.filter(el => el.gender === value);
+  }
+
```
Broken syntax. Avoid such mistakes :)

Add brackets and use prettier

zonzujiro commented on submissions/SamVal007/Friends_APP/script.js:
```diff
@@ -0,0 +1,125 @@
+};
+
+document.addEventListener('DOMContentLoaded', function () {
+  makeFriendList()
+    .then(() => {
+      displayCards();   
+    });
```
```suggestion
  makeFriendList().then(displayCards);
```

zonzujiro commented on submissions/SamVal007/Friends_APP/script.js:
```diff
@@ -0,0 +1,125 @@
+    });
+
+  SEARCH_FIELD.addEventListener('input', function () {
+    let searchStr = '';
```
No need in `Str` in variable. Type of data is redundant info in variable name. Please, find another one.

zonzujiro commented on submissions/SamVal007/Friends_APP/script.js:
```diff
@@ -0,0 +1,125 @@
+  MENU_FORMS.addEventListener('change', ({target: radioButton}) => {
+    putPersonsInDisplayList();
+
+    if (radioButton.name === 'genderField') filtrByGender(radioButton.value);
```
Sometimes you use brackets, sometimes don't.

You should pick one approach and use it :)

zonzujiro commented on submissions/SamVal007/Friends_APP/script.js:
```diff
@@ -0,0 +1,125 @@
+
+    if (radioButton.name === 'genderField') filtrByGender(radioButton.value);
+
+    if (radioButton.name === "nameField") {
+      nameSorters[radioButton.value]();
+    }
+
+    if (radioButton.name === "ageField") {
+      ageSorters[radioButton.value]();
+    }
```
`ageSorters[radioButton.name]()`?

zonzujiro commented on submissions/SamVal007/Friends_APP/script.js:
```diff
@@ -0,0 +1,125 @@
+
+const ageSorters = {
+  ageLow: () => persons.sort((a, b) => compareAge(b, a)),
+
```
Redundant empty line

zonzujiro reviewed (changes requested):
You should pay more attention to code style :) Redundant empty lines, missing curly brackets, etc.
